# Bacterial Genomics
## Monday Viikki C building, lecture room 130 
| Time | Description|
| --- | --- |
| 09:00-09:30 | **Jenni Hultman and Mirko Rossi**: Introduction (including general structure of the course and group work)|
| 09:30-09:45 | **Mirko Rossi**: Basic usage of cPouta cloud computing service at CSC |
| 09:45-10:45 | **Mirko Rossi**: [Prepare the Virtual Machine](https://github.com/INNUENDOCON/MicrobialGenomeMetagenomeCourse/blob/master/MPM_starting_VM.md) and introduction of docker|
| 10:45-11:00 | *break* - Note: the coffee break is not included in the course |
| 11:00-11:45 | **João Carriço**: Making Use of NGS Data: From Reads to Trees and Annotations |
| 11:45-12:00 | **Mirko Rossi**: *INSTRUCTION FOR GROUP WORK* |
| 12:00-13:30 | **Working with your data** and **lunch** |
| 13:30-17:00 | **Miguel Machado**: [From reads to assemblies: working with INNUca pipeline](https://github.com/INNUENDOCON/MicrobialGenomeMetagenomeCourse/blob/master/MPM_workingwithINNUCA.md) |

## Tuesday Biocentre 3, lecture room 1402
| Time | Description |
| --- | --- |
| 08:30-11:45 | **Miguel Machado**: *In silico* typing using ReMatCh and Abricate |
| 11:45-12:00 | **Mirko Rossi**: *INSTRUCTION FOR GROUP WORK* |
| 12:00-13:30 | **Working with your data** and **lunch** |
| 13:30-16:00 | **Mickael Silva**: Whole genome and core genome MLST analysis using *chewBBACA* suite (see also [*chewBBACA wiki*](https://github.com/B-UMMI/chewBBACA/wiki)) |
| 16:00-16:10 | **Mirko Rossi**:*INSTRUCTION FOR GROUP WORK* |
| 16:10-17:00 | **Working with your data** 

## Wednesday Viikki C building, lecture room 130 
| Time | Description |
| --- | --- |
| 08:30-10:45 | **Miguel Machado & Bruno Gonçalves**: Annotation (*PROKKA*), pangenome anaylysis (*ROARY*) and GWAS (*SCOARY*)|
| 10:45-11:00 | **Mirko Rossi**: *INSTRUCTION FOR GROUP WORK* |
| 11:00-12:30 | **Working with your data** and **lunch** |
| 12:30-14:30 | **Bruno Gonçalves**: Integrating phylogenetic inference and metadata visualization for NGS data |
| 14:30-17:00 | **METAGENOMIC** |
| 14:30-15:30 | **Antti Karkman** Data download, QC (*FASTQC*) and trimming (*CUTADAPT*)|
| 15:30-17:00 | **Jenni Hultman** Assembly (*MEGAHIT*) and installation of programs |

## Thusday Lab building, lecture room 228
| Time | Description |
| --- | --- |
| 08:00-10:00 | **Jenni Hultman**: Assembly check and mapping with (*BOWTIE2*) |
| 10:00-12:00 | **Antti Karkman**: Read based analysis for 16S rRNA (*Metaxa2*) and ARG (*ResFinder*) |
| 12:00-14:00 | **Working with your data** and **lunch** |
| 14:00-17:00 | **Antti Karkman and Jenni Hultman** Prepping files for Anvio |


## Friday Viikki C building, lecture room 130 
| Time | Description |
| --- | --- |
| 09:00-10:00 | **Tom Delmont**: Introduction to Anvio NB. Room 2041 Biocenter 2 |
| 10:30-12:00 | **Tom Delmont**: Metagenome visualization in Anvio (*ANVIO*) |
| 12:00-13:00 | **Working with your data** and **lunch** |
| 13:00-16:00 | **Tom Delmont** Construction of MAGs with Anvio |
| 16:00-17:00 | **Jenni Hultman** Wrapping up |


