# From microbial genomics to metagenomics

# FEEDBACK
The summary of the feedback questionnaire is available from this [link](https://github.com/INNUENDOCON/MicrobialGenomeMetagenomeCourse/blob/master/Statistics_Export_MGMC_2017_summaryPdf.pdf). All the answers can be downloaded from this [link](https://github.com/INNUENDOCON/MicrobialGenomeMetagenomeCourse/blob/master/Content_Export_MGMC_2017_All.feedback.zip). Response rate was **67%** and **> 91%** of the students rate the course **good or very good**. Thanks to the teachers and the students for this nice intense week.  

## General introduction
Microbial community structure and diversity, and population structure are fundamental aspects to understand evolution, niche adaptation and demographic history of bacterial species. As NGS sequencing has become cost effective and accessible, sequencing populations of bacteria across the whole genome provides unprecedented resolution to investigate within-host evolution, transmission history, and population structure. Moreover, analysis of genetic content of microbial communities through metagenomics has become the mainstream methodology. During the one week course the students will learn and apply bioinformatic techniques to perform population genetics and study microbial communities with metagenomic approaches. The goal is to become familiar with the bioinformatic analysis tools and to be able to utilize them in own research after the course.

**Everything is possible thanks the support of CSC**

## [Schedule](https://github.com/INNUENDOCON/MicrobialGenomeMetagenomeCourse/edit/master/ScheduleofCourse.md)
The course will be held at the Faculty of Veterinary Medicine, University of Helsinki. 
* **Monday to Wednesday morning**: [bacterial genomics](https://github.com/INNUENDOCON/MicrobialGenomeMetagenomeCourse/blob/master/Genomics.md) 
* **Wednesday afternoon to Friday**: [metagenomics](https://github.com/INNUENDOCON/MicrobialGenomeMetagenomeCourse/blob/master/Metagenomics.md)
## Place
* Monday, Wednesday & Friday Viikki C building, lecture room 130 (Latokartanonkaari 5)
* Tuesday Biocentre 3, lecture room 1402 (Viikinkaari 1)
* Thursday Lab building, lecture room 228 (Koetilantie 5)
## Target group
This course is targeted at PhD and MSc students who are interested in performing bacterial population analysis using whole-genome sequencing and/or using metagenomics for investigating microbial communities.
## Learning outcomes
Microbial genomics: QA/QC of the raw sequence data, cleaning and assembly; Gene-by-gene annotation; Genealogy reconstruction and population structure analysis; Visualization of data; GWAS and comparative analysis; Microbial metagenomics: quality control, filtering and assembly to taxonomic classification, clustering, functional assignment, analyses of microbial community composition and comparative metagenomics
## Internet connection 
Inside the class room you can find both eduroam and Univ. Helsinki HUPnet. Instructions on how to connect in HelpDesk website
## Organizers
* Docent Jenni Hultman, University of Helsinki, Finland 
* Associate Professor Mirko Rossi, University of Helsinki, Finland
## Teachers
* Docent Jenni Hultman, University of Helsinki, Finland
* Associate Professor Mirko Rossi, University of Helsinki, Finland
* Dr João Carriço, University of Lisbon, Portugal
* Bruno Gonçalves, University of Lisbon, Portugal
* Miguel Machado, University of Lisbon, Portugal
* Mickael Silva, University of Lisbon, Portugal 
* Dr Antti Karkman, University of Gothenburg, Sweden
* Dr Tom Delmont, University of Chigaco, USA
